From 28ef90a0f71372040ba11ad0a917b2bec88bc17e Mon Sep 17 00:00:00 2001
From: Amir Vadai <avadai@redhat.com>
Date: Thu, 11 Sep 2014 12:02:11 -0400
Subject: [ethernet] mlx4: Avoid dealing with MAC index in UPDATE_QP wrapper if not needed

Message-id: <f4daba02b0878f0db2830e3653b50c92cebccdf9.1410425016.git.avadai@redhat.com>
Patchwork-id: 93699
O-Subject: [Patch RHEL7.1 V1 195/197] net/mlx4: Avoid dealing with MAC index in UPDATE_QP wrapper if not needed
Bugzilla: 1107617 1107618 1107619 1107620
RH-Acked-by: Jes Sorensen <Jes.Sorensen@redhat.com>

From: Matan Barak <matanb@mellanox.com>

The current wrapper implementation of the UPDATE_QP command tries to get
the MAC index, even if MAC wasn't set by the VF. Fix it up to only handle
the MAC field if it's valid.

Fixes: ce8d9e0 ('net/mlx4_core: Add UPDATE_QP SRIOV wrapper support')
Signed-off-by: Matan Barak <matanb@mellanox.com>
Signed-off-by: Or Gerlitz <ogerlitz@mellanox.com>
Signed-off-by: David S. Miller <davem@davemloft.net>
(cherry picked from commit b7834758e6f33da401f584035f1b6b62d0672aab)
Signed-off-by: Amir Vadai <avadai@redhat.com>
---

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/drivers/net/ethernet/mellanox/mlx4/resource_tracker.c b/drivers/net/ethernet/mellanox/mlx4/resource_tracker.c
index f87a4a3..2fe61b6 100644
--- a/drivers/net/ethernet/mellanox/mlx4/resource_tracker.c
+++ b/drivers/net/ethernet/mellanox/mlx4/resource_tracker.c
@@ -3998,13 +3998,17 @@ int mlx4_UPDATE_QP_wrapper(struct mlx4_dev *dev, int slave,
  }
 
  port = (rqp->sched_queue >> 6 & 1) + 1;
- smac_index = cmd->qp_context.pri_path.grh_mylmc;
- err = mac_find_smac_ix_in_slave(dev, slave, port,
-     smac_index, &mac);
- if (err) {
-  mlx4_err(dev, "Failed to update qpn 0x%x, MAC is invalid. smac_ix: %d\n",
-    qpn, smac_index);
-  goto err_mac;
+
+ if (pri_addr_path_mask & (1ULL << MLX4_UPD_QP_PATH_MASK_MAC_INDEX)) {
+  smac_index = cmd->qp_context.pri_path.grh_mylmc;
+  err = mac_find_smac_ix_in_slave(dev, slave, port,
+      smac_index, &mac);
+
+  if (err) {
+   mlx4_err(dev, "Failed to update qpn 0x%x, MAC is invalid. smac_ix: %d\n",
+     qpn, smac_index);
+   goto err_mac;
+  }
  }
 
  err = mlx4_cmd(dev, inbox->dma,
-- 
1.7.1